### ISSUE TITLE
`import M.{this}` is a no-op?

### ISSUE THREAD

**Issue by BryantLam** [chapel-lang/chapel#18267]
Bug. This code uses `A.fnA()` but it doesn't resolve because `import A.{this}` doesn't appear to do anything.

```chapel
module A {
    proc fnA() {}
    module B {
        proc fnB() {}
    }
}

module SomeOtherFile {
    proc main() {
        // import A;
        // import A.B;
        import A.{this, B};
        A.fnA();
        B.fnB();
    }
}
```

```terminal
chpl version 1.25.0 pre-release (aa8f9f979)
+ chpl -o this-path this-path.chpl
this-path.chpl:9: In function 'main':
this-path.chpl:13: error: 'A' undeclared (first use this function)
```

Replacing with the commented-out lines is the workaround. (I believe the statements are equivalent.)

**Comment by lydia-duncan** [comment]
I don't recall `this` being specified as something that would enable importing the module name itself and can't seem to find where in the docs that would be suggested, so far as I know we don't currently have a way to specify bringing in both the module name and symbols within the module via an import statement.  This seems like a feature request rather than a bug

**Comment by BryantLam** [comment]
If not supported, this should generate an error. I thought the form would work because it was syntactically accepted. I could have `import A, A.B` instead; only annoying with repetition if `A` is a long symbol name. I don't need it to be a feature request, though it probably isn't that much harder to implement the feature after implementing the error message.

**Comment by bradcray** [comment]
I agree that if it's not intended to be accepted, it should generate an error (and doubt Lydia would disagree).  This caused me to sketch out the following code to see what happens, which was surprising in its seeming inconsistency to me:

```chapel
module M {
  proc foo() {
  }
}

module N {
  import M./* something here */;
  proc main() {
  }
}
```

Specifically if "/* something here */" is replaced by:

* `bar`: we get `error: Cannot find symbol 'bar' in module 'M'`
* `this`: we get `error: Cannot find symbol 'this' in module 'M'`
* `{bar}`: we get `error: Bad identifier, no known 'bar' defined in 'M'`
* `{this}`: things work

So there's obviously an asymmetry here, both in terms of the error message wording and the behavior.  @lydia-duncan is on vacation this week, but I expect she'll be able to explain why this would be when she returns.

**Comment by lydia-duncan** [comment]
That's . . . odd, to say the least.  I'll look into that this week

**Comment by lydia-duncan** [comment]
> My original hypothesis was that "this" was getting matched to the `this` method on types so that you can access into the type with square brackets and parentheses, though that would be a bit of an issue since I thought limitation clauses didn't apply to method names directly (and also, in that code it shouldn't find it since it shouldn't go through the private use of ChapelStandard that is automatically inserted). It turns out that **is** why it thinks `import M.{this}` is okay.

Note that this check is only for if it's valid to name `this` in the import's list for unqualified access - using `this` in the scope with the import statement correctly determines that you can't find the symbol.  The import/use list not correctly recognizing that you can't actually reference `this` is another instance of #16982

**Comment by lydia-duncan** [comment]
@BryantLam - if we fix #16982, so `import M.{this}` only is accepted if M has a `this` method visible, does that address your concern?

**Comment by bradcray** [comment]
> follows a slightly different resolution path

I don't quite follow why it would be at all different.  Are there cases where `import M.{x, y, z};` means something different than `import M.x; import M.y; import M.z;`?  (I thought "no"?).

> so import M.{this} only is accepted if M has a this method visible

I'm not Bryant, but that makes sense to me (though it raises the question whether a module should ever define a standalone `this` routine... but that's another topic).

**Comment by lydia-duncan** [comment]
> > follows a slightly different resolution path
> 
> I don't quite follow why it would be at all different. Are there cases where `import M.{x, y, z};` means something different than `import M.x; import M.y; import M.z;`? (I thought "no"?).

It has more to do with `import M.N.O.x` - we need `M`, `N`, and `O` to be modules and for the last symbol after a `.` to always be valid (otherwise the statement itself is completely useless).  That necessitates a different resolution path than the names in the `{}` part of an import statement - for the `{}` names, we know the starting point module, we've resolved all the `.` parts prior to it.  It's possible for there to be modules listed in `{ }` but it's not essential that they are modules - it is essential that every symbol prior to the `.{` is a module, though.  At the end, `import M.x` should behave equivalently to `import M.{x}` for a user, and you want the implementation to be as unified as possible, but there's always a slight point of difference.

**Comment by BryantLam** [comment]
My question is probably too far in the weeds, but why wouldn't `import M.{x, y, z}` take an implementation path to desugar as (1) `import M.{x}, <assume M resolves>.{y}, <assume M resolves>.{z}` or (2) `import M.<for all item in { ... }>`?

Then the `import M.x` behavior would desugar to `import M.{x}` and there's no difference in resolution paths anymore?


> @BryantLam - if we fix #16982, so `import M.{this}` only is accepted if M has a `this` method visible, does that address your concern?

I didn't realize you could do that. You can define a `this` method in module scope and then function-call the module symbol in a different scope?

From a Rust background (`this == self`), my assumption was that these statements were equivalent:

- `import M.{this}` same as `import M`
- `import M.{this, A, B}` same as `import M; import M.A; import M.B`
- `import M.{this, <everything inside M>}` same as `use M`

But I didn't realize there was a special `this` method you could define in a module. This might be a design difference with Rust, but what would be the purpose to importing a `this` method? Would you expect a user that did `import M` to not be allowed to function-call on the module `M` unless `import M.this` was also executed?

**Comment by lydia-duncan** [comment]
> My question is probably too far in the weeds, but why wouldn't `import M.{x, y, z}` take an implementation path to desugar as (1) `import M.{x}, <assume M resolves>.{y}, <assume M resolves>.{z}` or (2) `import M.<for all item in { ... }>`?
> 
> Then the `import M.x` behavior would desugar to `import M.{x}` and there's no difference in resolution paths anymore?

We don't perform any transformations on `import M.x` until we know that `x` isn't a module, at which point we basically make it `import M.{x}`.  But we have to get to that point and that requires the check I'm describing.  I guess putting a converse spin on it, do you think `import M.{N}` should transform into `import M.N` when N is a module?

> 
> > @BryantLam - if we fix #16982, so `import M.{this}` only is accepted if M has a `this` method visible, does that address your concern?
> 
> I didn't realize you could do that. You can define a `this` method in module scope and then function-call the module symbol in a different scope?

You can define a `this` method on a type, enabling it to have at least some of its contents accessed using `[]` and `()` like arrays do.  It's visible in a module like other methods are.  Brad mentioned that we might consider making it available as a module standalone function, though I'm not sure what the use case would be and agree that we shouldn't get too distracted by it.  As an aside, there's a debate about if we want to rename the `this` method to something else over on #18263.

**Comment by BryantLam** [comment]
> I guess putting a converse spin on it, do you think `import M.{N}` should transform into `import M.N` when N is a module?

Regardless of forward or backward transformation, I would expect `import M.{N.{x, y}, L}` to be "same" as `import M.N.x, M.N.y, M.L`. where `M`, `N`, and `L` are modules. I think I see your point though: there's a different resolution path taken for module items vs. non-module items.
